### DISCUSSION TITLE
Memory usage on ra_xxx_log_ets tables

### DISCUSSION THREAD

**Discussion by albertored** [rabbitmq/ra#453] [Q&A]
We have an application that sometimes applies commands with a big payload in the raft machine.
In such situations we get a lot of memory "wasted" in the raft machine process itself but we are able to solve this by returning the `:garbage_collect` effect. We still have some ETS tables (owned by `ra_xxx_log_ets`) that consume a lot of memory and it is not clear how/when they are freed or if there are parameters to configure this behaviour.
Do you have some advice on how to reduce this memory consumption?
Thanks

**Comment by Kation** [comment]
Same problem.  
We have a message delive in dead loop because a bug that throw exception with manual ack.  
Then ra_log_ets size increase very large.  
Even we handled this message for a long time, ra_log_ets never relase memory.

**Reply by michaelklishin** [reply]
I don't know what a "dead loop" is. We are not aware of any bugs in RabbitMQ where it would "throw exception with manual ack".

What is certain is that with workloads where consumers do not acknowledge messages, the Raft log of the quorum queue they consume from **will not** be truncated until the oldest message is acknowledged. This has been a quorum queue behavior since 2018, and why [this delivery timeout](https://www.rabbitmq.com/docs/consumers#acknowledgement-timeout) is enforced on consumers.

If you do not acknowledge messages, your data set won't be truncated, that's an oversimplified but nonetheless correct explanation of quorum queue behavior.

That's all that can be said without having an executable way to reproduce the behavior of your apps. We do not guess in this community.

**Reply by michaelklishin** [reply]
Quorum queues have another related protection mechanism against consumers that get a delivery, fail to acknowledge it and either requeue it or fail (which causes the delivery to be requeued): [Poison Message Handling](https://www.rabbitmq.com/docs/quorum-queues#poison-message-handling) a.k.a. the limit on the number of redeliveries.

In 4.0, there will be a [default limit](https://github.com/rabbitmq/rabbitmq-server/pull/11937) because RabbitMQ gets the blame for faulty applications that go into these "delivery-failure" loops.

**Reply by Kation** [reply]
We are warpping the message handler that exception throw by business code will use `BasicNack`.  
So that message will deliver again and again.

When we found there a business code bug case by invalid message body, we stop application and purge messages in this quorum queue.  
And ra_log_ets does not release.

**Reply by michaelklishin** [reply]
@Kation requeueing a message over and over is the same as not acknowledging it at all, and exactly why https://www.rabbitmq.com/docs/quorum-queues#poison-message-handling exists. We need an executable way to reproduce. We do not guess in this community.

**Reply by michaelklishin** [reply]
In fact, it is extremely likely to be a classic consumer "crash-requeue-crash-requeue" scenario and nothing else. Compaction will be triggered as it should be when the crash loop is eliminated. But let's see what the executable way to reproduce will look like.

**Reply by Kation** [reply]
The point is that message maybe importance and should consume after consumer application update.   
I will try to write a demo later.

**Reply by michaelklishin** [reply]
I don't see why an important message must be requeued may times over and over. **That** is the problematic scenario. Quorum queues design assumes that messages are acknowledged reasonably quickly, not after a 50 redelivery loop one day later.
